rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+}
+
+#[derive(Debug, Error)]
+enum RpcError {
+    #[error("RPC call failed with error: {0}")]
+    RpcCall(String),
+    #[error("Request failed with error: {0}")]
+    Request(String),
+    #[error("Failed to cast from: {0} to: {1} with error: {2}")]
+    Cast(String, String, String),
+}
+
+/// [`BlockValue`] is an Enum that represent which block we are going to use to retrieve information.
+#[allow(dead_code)]
+pub enum BlockValue {
+    /// String one of: ["latest", "pending"]
+    Tag(serde_json::Value),
+    /// Integer
+    Number(serde_json::Value),
+    /// String with format: 0x{felt252}
+    Hash(serde_json::Value),
+}
+
+impl BlockValue {
+    fn to_value(&self) -> serde_json::Value {
+        match self {
+            BlockValue::Tag(block_tag) => block_tag.clone(),
+            BlockValue::Number(block_number) => json!({ "block_number": block_number }),
+            BlockValue::Hash(block_hash) => json!({ "block_hash": block_hash }),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct RpcResponseContractClass {
+    result: SNContractClass,
+}
+
+// We use this new struct to cast the string that contains a [`StarkFelt`] in hex to a [`StarkFelt`]
+struct StarkFeltHex;
+
+impl<'de> DeserializeAs<'de, StarkFelt> for StarkFeltHex {
+    fn deserialize_as<D>(deserializer: D) -> Result<StarkFelt, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: PrefixedBytesAsHex<32> = PrefixedBytesAsHex::deserialize(deserializer)?;
+        Ok(bytes.try_into().unwrap())
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Deserialize)]
+struct RpcResponseFelt {
+    #[serde_as(as = "StarkFeltHex")]
+    result: StarkFelt,
+}
+
+pub struct RpcBlockInfo {
+    /// The sequence number of the last block created.
+    pub block_number: BlockNumber,
+    /// Timestamp of the beginning of the last block creation attempt.
+    pub block_timestamp: BlockTimestamp,
+    /// The sequencer address of this block.
+    pub sequencer_address: ContractAddress,
+}
+
+type RpcResponse = ureq::Response;
+
+impl RpcState {
+    pub fn new(chain: RpcChain, block: BlockValue) -> Self {
+        if env::var("INFURA_API_KEY").is_err() {
+            dotenv().expect("Missing .env file");
+        }
+        Self {
+            chain,
+            api_key: env::var("INFURA_API_KEY")
+                .expect("Missing API Key in environment: INFURA_API_KEY"),
+            block,
+        }
+    }
+
+    fn rpc_call<T: for<'a> Deserialize<'a>>(
+        &self,
+        params: &serde_json::Value,
+    ) -> Result<T, RpcError> {
+        Self::deserialize_call(self.rpc_call_no_deserialize(params)?)
+    }
+
+    fn rpc_call_no_deserialize(&self, params: &serde_json::Value) -> Result<RpcResponse, RpcError> {
+        ureq::post(&format!(
+            "https://{}.infura.io/v3/{}",
+            self.chain, self.api_key
+        ))
+        .set("Content-Type", "application/json")
+        .set("accept", "application/json")
+        .send_json(params)
```

Comment:
I'm not sure if it's intended here, but I remember changing the JSON serialization to raw `format!()` strings.